### ISSUE TITLE
Akka 2.5.14 compilation broken with custom-placed JDK

### ISSUE THREAD

**Issue by GrafDiffusore** [akka/akka#25423]
Hi,

In relation to #25302 -- it might be this change broke compilation of Akka for me -- my JDK is manually downloaded and put into `/home/user/bin` directory and added to `PATH` (and same goes for `scala` and `sbt` as well). Here's output of `ls -1 /home/user/bin`:
```
jdk1.8.0_162
jdk8 # symlink to above, used in PATH and JAVA_HOME
sbt   # symlink to below, used in PATH and SBT_HOME
sbt-1.1.6
scala   # symlink to below, used in PATH and SCALA_HOME
scala-2.12.6
```
Is it possible to run, e.g.,  `java -v` or `scala -version` process to discover version?

Partial output from `sbt compile` of Akka 2.5.14 using setup from bin dir above:
```
[info] Formatting 11 Scala sources ProjectRef(uri("file:/home/user/files/src/distributed/akka/"), "akka-distributed-data")(multi-jvm) ...
[error] java.util.NoSuchElementException: key not found: 8
[error] 	at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:98)
[error] 	at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:96)
[error] 	at akka.AkkaBuild$.$anonfun$defaultSettings$6(AkkaBuild.scala:104)
```

**Comment by eshepelyuk** [comment]
Same issue happens to me on Windows. I'm having JDK8 downloaded and installed in `d:\opt\jdk8` folder, and `JAVA_HOME` set to point into that location and `JAVA_HOME\bin` added to my `PATH` env variable .

When running a build I'm receiving following errors
https://gist.github.com/eshepelyuk/57d27959aa1a4e969ebb1a9cb75d01c1

From the same command line I succesfully run other `Gradle` or `Maven` builds.

I've been able to woraround the issue by patching `project/AkkaBuild` file by replacing `CrossJava.Keys.fullJavaHomes.value("8") + "/jre/lib/rt.jar"` to  `"d:/opt/jdk8/jre/lib/rt.jar"`

Check [this commit](https://github.com/eshepelyuk/akka/commit/404f01adc817d71158ad9009135c074a016801f3) for details.

**Comment by lefou** [comment]
Has anybody validated, that this issue is fixed by #25457 ?

I can't build master and see the same error, although #25457 is merged.

**Comment by lefou** [comment]
```
$ echo $JAVA_HOME
/home/lefou/.gentoo/java-config-2/current-user-vm

$ ls -lad /home/lefou/.gentoo/java-config-2/current-user-vm
lrwxrwxrwx 1 lefou lefou 26 22. Nov 09:20 /home/lefou/.gentoo/java-config-2/current-user-vm -> /usr/lib/jvm/icedtea-bin-8
$ ls -lad /usr/lib/jvm/icedtea-bin-8
lrwxrwxrwx 1 root root 22 22. Okt 22:58 /usr/lib/jvm/icedtea-bin-8 -> /opt/icedtea-bin-3.9.0

$ java -version
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (IcedTea 3.9.0) (Gentoo icedtea-3.9.0)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
```

Invoking with explicit `JAVA_HOME` doesn't help.

```
$ JAVA_HOME=/opt/icedtea-bin-3.9.0 sbt publishM2
...
[error] java.util.NoSuchElementException: key not found: 8
[error]         at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:98)
[error]         at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:96)
[error]         at akka.AkkaBuild$.$anonfun$defaultSettings$2(AkkaBuild.scala:102)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:40)
[error]         at sbt.std.Transform$$anon$4.work(System.scala:67)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:269)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error]         at sbt.Execute.work(Execute.scala:278)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:269)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
...
```

The "java" command is a shell script on PATH though. It essentially `exec`s the real java bin.

```
$ which java
/usr/bin/java
```

**Comment by lefou** [comment]
Same for "official" Oracle JDK.

**Comment by lefou** [comment]
What is so special in the akka build that it needs to assume/infer those versions from paths? Without thinking much about it in the akka context, parsing the output of `java -version` seems more reliable to me then parsing some path.

**Comment by lefou** [comment]
Also, the current running JVM typically sets a system property `java.version`.

**Comment by raboof** [comment]
> What is so special in the akka build that it needs to assume/infer those versions from paths?

We want Akka to be buildable with both JDK8 and JDK9+, and always produce a jar that is usable on Java 8 for both Scala 2.11 and 2.12 (we're working on 2.13 ;) ).

On JDK8, obviously the resulting jar will work on Java 8, though it will miss any features that rely on classes that are only available on Java 9 and later (specifically, the Reactive Streams support).

On JDK9+, in theory the resulting jar would work on Java 8 by simply passing `-release 8` to scalac and `--release 8` to javac. Unfortunately, `-release 8` is not available for the Scala 2.11 scalac, and `--release 8` restricts the use of `sun.misc.Unsafe` which we still use here and there.

For this reason, at least when building on JDK9+, we need to specifically set the Java bootclasspath to the one from the JDK8 installation.

When we drop support for 2.11 (not currently planned AFAIK) and remove our reliance on `sun.misc.Unsafe` on JDK9+ (some work done in this area) this can go away, but we'll be stuck with this for a while.

> JAVA_HOME=/opt/oracle-jdk-bin-1.8.0.181 sbt publishM2

Interesting, I haven't seen that naming convention before. At least that we can easily add, with a change like https://github.com/sbt/sbt/pull/4462

**Comment by raboof** [comment]
> Also, the current running JVM typically sets a system property `java.version`.

Right, indeed we could sidestep the whole issue when already on JDK8. Could you try if https://github.com/akka/akka/pull/25962 works for you?

**Comment by lefou** [comment]
Thanks for the explanations and the fix in #25962, which works for me.

**Comment by patriknw** [comment]
seems we can close this now
